miner/multi_collator.go

```diff
@@ -0,0 +1,394 @@
+	sideChainCh       chan types.Header
+	blockCollatorImpl BlockCollator
+
+	chainConfig *params.ChainConfig
+	chain       *core.BlockChain
+	pool        Pool
+}
+
+// each active collator runs mainLoop() in a goroutine.
+// It receives new work from the miner and listens for new blocks built from CollateBlock
+// calling Commit() on the provided collatorBlockState
+func (c *collator) mainLoop() {
+	for {
+		select {
+		case newWork := <-c.newWorkCh:
+			c.blockCollatorImpl.CollateBlock(&collatorBlockState{work: newWork, c: c, done: false}, c.pool)
+			// signal to the exitCh that the collator is done
+			// computing this work.
+			c.workResultCh <- collatorWork{env: nil, interrupt: nil, counter: newWork.counter}
+		case <-c.exitCh:
+			// TODO any cleanup needed?
+			return
+		case newHead := <-c.newHeadCh:
+			// TODO call hook here
+			_ = newHead
+		case sideHeader := <-c.sideChainCh:
+			_ = sideHeader
+			// TODO call hook here
+		default:
+		}
+	}
+}
+
+var (
+	// collator
+	workResultChSize = 10
+	newWorkChSize    = 10
+	newHeadChSize    = 10
+	sideChainChSize  = 10
+)
+
+// MultiCollator manages multiple active collators
+type MultiCollator struct {
+	counter                 uint64
+	responsiveCollatorCount int
+	collators               []collator
+	pool                    Pool
+	interrupt               *int32
+}
+
+func NewMultiCollator(chainConfig *params.ChainConfig, chain *core.BlockChain, pool Pool, strategies []BlockCollator) MultiCollator {
+	collators := []collator{}
+	for _, s := range strategies {
+		collators = append(collators, collator{
+			newWorkCh:         make(chan collatorWork, newWorkChSize),
+			workResultCh:      make(chan collatorWork, workResultChSize),
+			exitCh:            make(chan struct{}),
+			newHeadCh:         make(chan types.Header, newHeadChSize),
+			blockCollatorImpl: s,
+			chainConfig:       chainConfig,
+			chain:             chain,
+			pool:              pool,
+		})
+	}
+	return MultiCollator{
+		counter:                 0,
+		responsiveCollatorCount: 0,
+		collators:               collators,
+		interrupt:               nil,
+	}
+}
+
+func (m *MultiCollator) Start() {
+	for _, c := range m.collators {
+		go c.mainLoop()
+	}
+}
+
+func (m *MultiCollator) Close() {
+	for _, c := range m.collators {
+		select {
+		case c.exitCh <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// SuggestBlock sends a new empty block to each active collator.
+// collators whose receiving channels are full are noted as "unresponsive"
+// for the purpose of not expecting a response back (for this round) during
+// polling performed by Collect
+func (m *MultiCollator) SuggestBlock(work *environment, interrupt *int32) {
+	if m.counter == Uint64Max {
+		m.counter = 0
+	} else {
+		m.counter++
+	}
+	m.responsiveCollatorCount = 0
+	m.interrupt = interrupt
+	for _, c := range m.collators {
+		select {
+		case c.newWorkCh <- collatorWork{env: work.copy(), counter: m.counter, interrupt: interrupt}:
+			m.responsiveCollatorCount++
+		}
```

Comment:
This is a select with only one case -- I think it's equivalent to:
```suggestion
		c.newWorkCh <- collatorWork{env: work.copy(), counter: m.counter, interrupt: interrupt}
		m.responsiveCollatorCount++
		
```
Is that the intent? Or did you mean to have  a default-case too?

Reply:
Yeah. Missing a default case (shouldn't block when `newWorkCh` is full).